### PR TITLE
Add API store to classifier

### PR DESCRIPTION
**Pull request by rogerhutchings**
Package: lib-classifier

Adds a store for env settings in the classifier

### CODE REVIEW

**Comment by srallen** [comment]
I'm happy to put together a quick branch to show what I mean.

**Comment by rogerhutchings** [comment]
It would be a band-aid; it doesn't solve the real problem, which is that the endpoint should be being set by the consuming app. Aside from fixing this issue, it's something we'd also need for e.g. deploying staging branches and 1715 Labs work.

It wouldn't -- shouldn't -- be hardcoded in any case, it should be set in the consuming app by an environment variable; which is basically what webpack would be doing anyway (it does a find and replace under the hood).

**Comment by rogerhutchings** [comment]
If anything, actually, I'd advocate moving that config out to its own package. The client should be determining how we interact with the API, not the target endpoint.

File: packages/lib-classifier/src/store/ClassificationStore.js [hunk 1]
```diff
@@ -168,7 +169,7 @@ const ClassificationStore = types
         self.classificationQueue.add(classification)
       } else {
         try {
-          const response = yield client.post(`/${self.type}`, { classifications: classification })
+          const response = yield client.post(`/${self.type}`, { classifications: classification }, undefined, host)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/ClassificationStore.js]: This saves all classifications as anonymous.

File: packages/lib-classifier/src/store/ResourceStore.js [hunk 2]
```diff
@@ -14,10 +14,11 @@ const ResourceStore = types
   .actions(self => ({
     fetchResource: flow(function * fetchResource (id) {
       const client = getRoot(self).client.panoptes
+      const host = getRoot(self).apiHost
       const { type } = self
       self.loadingState = asyncStates.loading
       try {
-        const response = yield client.get(`/${type}/${id}`)
+        const response = yield client.get(`/${type}/${id}`, undefined, undefined, host)
```
Comment (eatyourgreens) [inline packages/lib-classifier/src/store/ResourceStore.js]: This is missing an auth token for protected resources too.

**Comment by eatyourgreens** [comment]
Changes so far seem to be removing auth tokens and query params from requests. I'm in agreement with @srallen here. This is introducing too much coupling between the classifier and the API to be comfortable for me.

File: packages/lib-classifier/src/store/ClassificationStore.js (hunk 1 above)
Comment (rogerhutchings) [reply packages/lib-classifier/src/store/ClassificationStore.js]: Auth hasn't been hooked up in the classifier AFAIK

File: packages/lib-classifier/src/store/ResourceStore.js (hunk 2 above)
Comment (rogerhutchings) [reply packages/lib-classifier/src/store/ResourceStore.js]: Auth hasn't been hooked up in the classifier AFAIK

File: packages/lib-classifier/src/store/ClassificationStore.js (hunk 1 above)
Comment (rogerhutchings) [reply packages/lib-classifier/src/store/ClassificationStore.js]: They already are, if there was no authorisation argument being passed in

File: packages/lib-classifier/src/store/ResourceStore.js (hunk 2 above)
Comment (rogerhutchings) [reply packages/lib-classifier/src/store/ResourceStore.js]: Same again

**Comment by rogerhutchings** [comment]
There aren't any auth tokens or query params being passed in at this point - all I've done is add an extra argument specifying the endpoint.

**Comment by srallen** [comment]
> it's something we'd also need for e.g. deploying staging branches (since that would involve a production build against the staging endpoint)

I'm confused about how my suggested solution wouldn't fix this? We would just use the build dev script and webpack config for those deploys.

> 1715 Labs work (which will most likely use a separate API)
Do we know this? I think we're getting ahead of ourselves a bit. We just need to fix the environment for our own setup right now.

I don't think having an entire store is quite the solution here and seems a bit heavy handed. What I'd probably suggest instead is both the npm script/webpack config variant for the immediate fix, then a longer term fix (to support different host endpoints) would be to add a basic class initializer for panoptes.js that support and stores basic configs like this rather than a separate package. It'd be the one thing I think it makes sense as a class for the client. Then the consumer can configure and initialize as needed.

**Comment by rogerhutchings** [comment]
I've since changed it keep the endpoint in the store context, which is what we do for the auth / panoptes clients: https://github.com/zooniverse/front-end-monorepo/pull/499/files#diff-a8a199f3b84c95f0423407dcc326b419R37

The reason I don't like the solution you proposed is because it spreads out config to another place. If you need to change an app-wide setting, you shouldn't have to do it in the app _and_ have to rebuild the classifier with a different build file -- the classifier component can be purely deterministic this way.

**Comment by eatyourgreens** [comment]
Working on the favourites button has reminded me that some of the API helpers, like collections and (I think) projects don't even let you specify the host anyway. You can only pass in the request payload and auth header.

**Comment by rogerhutchings** [comment]
Yeah, I noticed that - that seems like a missing feature from panoptes-js, given that the main request functions let you specify it.

**Comment by srallen** [comment]
> If you need to change an app-wide setting, you shouldn't have to do it in the app _and_ have to rebuild the classifier with a different build file

Oh, I don't mean doing both. I mean using webpack for the initial fix since that would be quick to do now and then later switch to a configuration for panoptes.js on a class instantiated by the consumer.

**Comment by eatyourgreens** [comment]
>Yeah, I noticed that - that seems like a missing feature from panoptes-js, given that the main request functions let you specify it.

I think it's by design. You don't want to be allowing the code to call one method with the staging host and a different method with the production host in the same session. I think the consuming app should specify once whether it's using the staging API or the production API and then the client uses the appropriate endpoints from that point onwards.

I really dislike how this PR is adding the host to every request method in the classifier app. I'm also not sure, given how short time is for TESS, whether this issue falls into fixing things that aren't really broken, and isn't really a priority.

**Comment by rogerhutchings** [comment]
It's fixing #444, which is fairly important. But fine.

**Comment by srallen** [comment]
Yes, it's by design. The basic HTTP request helpers can take differing hosts since my thoughts that if you were using the low level request function that you might want to do something really different(?), but the resource specific ones should use the host defined once by the process environment variable or by query param. 

panoptes.js should be better configurable for different hosts, as I think it was rightly pointed out 1715 labs might use something different, but we don't seem to agree yet on a solution and I also agree it's not necessary for TESS. The immediate solution is to have the webpack config correctly use the environment we want for development.

**Comment by eatyourgreens** [comment]
The 'Your Progress' component might be an example of using those low-level methods, where it wants to post a query to the stats service host: https://github.com/zooniverse/front-end-monorepo/blob/08183c5f92f3882daaca5b4eb6b73df5a84006a6/packages/app-project/stores/YourStats.js#L63-L75

**Pull request closed without merging**